### ISSUE TITLE
Java and Scala compilers treat records differently, Scala compiler not exposing accessor methods

### ISSUE THREAD

**Issue by ek-gh** [lampepfl/dotty#19386]
## Compiler version
3.3.1

## Minimized code
MyRecord.java: 
```
public record MyJavaRecord(String name, Integer age) {}
```
MyTest.scala
```
object MyScalaObject {
  @main def test(): Unit = {
    val javaRecord = new MyJavaRecord("Name", 50)
    println(javaRecord.name())
    println(javaRecord.age)
  }
}
```

If MyJavaRecord is compiled with java compiler, and MyScalaObject is compiled with scala compiler using class file generated by Java compiler for the record, you'd get no syntax errors from scala compiler.

If MyJavaRecord & MyScalaObject are both compiled with scala compiler, first println would give you syntax error.

## Expectation
When record is compiled via Scala compiler, it exposes accessor methods to the fields just like Java compiler.  We should get no errors compiling both with scala compiler.  Scala compiler should not enforce a syntax for java data type that is illegal by java compiler.

**Comment by som-snytt** [comment]
Same in Scala 2. Except more carets in Scala 3.
```
➜  i19386 scalac -d /tmp/sandbox *a
MyScalaObject.scala:5: error: not enough arguments for method apply: (i: Int): Char in class StringOps.
Unspecified value parameter i.
    println(javaRecord.name())
                           ^
1 error
➜  i19386 ~/projects/dotty/bin/scalac -d /tmp/sandbox *a
-- [E171] Type Error: MyScalaObject.scala:5:27 -------------------------------------------------------------------------
5 |    println(javaRecord.name())
  |            ^^^^^^^^^^^^^^^^^
  |            missing argument for parameter i of method apply in class StringOps: (i: Int): Char
1 error found
```

**Comment by ek-gh** [comment]
Folks, is there a plan to incorporate the fix to this issue into the 3.3 LTS?

**Comment by He-Pin** [comment]
I think this should be backported, and I'm using Java 21 LTS at work too.

**Comment by som-snytt** [comment]
No one asked, but https://github.com/scala/scala/pull/10715

**Comment by ek-gh** [comment]
IMHO incompatibility with Java is a significant bug to remain in LTS version of Scala.

**Comment by ek-gh** [comment]
Moreover, it represents a real challenge to developers who work on mix Java-Scala projects.

**Comment by som-snytt** [comment]
The workaround is `compileOrder := CompileOrder.JavaThenScala` per https://www.scala-sbt.org/1.x/docs/Java-Sources.html

Someone always mentions that setting eventually. I've never actually used it but it's an easy internet search.

**Comment by ek-gh** [comment]
Would this be required for tools such as Scala plugin for IntelliJ?  Or would this setting be integral to Scala 3.3 LTS compiler settings so no other client code / tool has to think about this?

**Comment by ek-gh** [comment]
Folks, 3.4.1 seems to have resolved this issue.  But since 3.3 is supposed to be Long Term Support, I will please request the fixes to be brought in to 3.3 as well.  As of 3.3.3, I am still seeing the same problem.  I am not sure who to reach for this request, so just adding it to this ticket.
